### PR TITLE
feat(platform): better contract state transitions

### PR DESCRIPTION
**Pull request by QuantumExplorer**
## Issue being fixed or feature implemented                                                                                                                                                                        
                                                                                                                                                                                                                  
  Implements V1 versions of DataContractCreateTransition and DataContractUpdateTransition that store contract fields directly rather than embedding the full DataContractInSerializationFormat. This is part of   
  Protocol Version 12.                                                                                                                                                                                            
                                                                                                                                                                                                                  
  Key problems solved:                                                                                                                                                                                            
  - V0 update transitions embed the entire contract even for small changes, wasting bandwidth and storage                                                                                                         
  - V1 update transitions now use delta-based updates, only transmitting what changed                                                                                                                             
  - V1 create transitions store fields directly instead of embedding a serialization format                                                                                                                       
                                                                                                                                                                                                                  
## What was done?                                                                                                                                                                                                  
                                                                                                                                                                                                                  
### New State Transition Versions                                                                                                                                                                                   
                                                                                                                                                                                                                  
  - DataContractCreateTransitionV1: Stores contract fields directly (owner_id, config, schema_defs, document_schemas, groups, tokens, keywords, description) rather than embedding                                
  DataContractInSerializationFormat                                                                                                                                                                               
  - DataContractUpdateTransitionV1: Delta-based updates with fields for:                                                                                                                                          
    - updated_schema_defs / new_schema_defs                                                                                                                                                                       
    - updated_document_schemas / new_document_schemas                                                                                                                                                             
    - new_groups, new_tokens                                                                                                                                                                                      
    - add_keywords / remove_keywords                                                                                                                                                                              
    - update_description                                                                                                                                                                                          
                                                                                                                                                                                                                  
### New Methods                                                                                                                                                                                                     
                                                                                                                                                                                                                  
  - DataContractUpdateTransition::from_contract_update(old_contract, new_contract, nonce, platform_version) - Computes delta between old and new contracts for V1 transitions                                     
  - DataContractUpdateTransition::from_data_contract_v0() - Explicit V0 transition creation for backwards compatibility                                                                                           
  - DataContract::apply_update() - Applies V1 delta updates to an existing contract                                                                                                                               
                                                                                                                                                                                                                  
### New Validation                                                                                                                                                                                                  
                                                                                                                                                                                                                  
  - basic_structure_v1 validation for V1 update transitions                                                                                                                                                       
  - transform_into_action_v1 for processing V1 transitions                                                                                                                                                        
  - New consensus errors: DataContractUpdateTransitionConflictingKeywordError, DataContractUpdateTransitionOverlappingFieldsError                                                                                 
                                                                                                                                                                                                                  
### Platform Version Updates                                                                                                                                                                                        
                                                                                                                                                                                                                  
  - Added DRIVE_ABCI_VALIDATION_VERSIONS_V8 with transform_into_action: 1                                                                                                                                         
  - Protocol Version 12 now defaults to V1 transitions (default_current_version: 1)                                                                                                                               
                                                                                                                                                                                                                  
## How Has This Been Tested?                                                                                                                                                                                       
                                                                                                                                                                                                                  
  - Unit tests: 26 v0_tests, 10 v1_tests for data contract update validation                                                                                                                                      
  - Integration tests: 5 data_contract_update_check_tx tests including:                                                                                                                                           
    - data_contract_update_check_tx_first_protocol_version (V0)                                                                                                                                                   
    - data_contract_update_check_tx_protocol_version_11 (V0, fee: 27002504030)                                                                                                                                    
    - data_contract_update_check_tx_latest_protocol_version (V1, fee: 27002430530)                                                                                                                                
    - Invalid update tests for both protocol versions                                                                                                                                                             
  - Strategy tests: run_chain_insert_one_new_identity_and_a_contract_with_updates                                                                                                                                 
  - DPP tests: Serialization and basic property tests                                                                                                                                                             
                                                                                                                                                                                                                  
## Breaking Changes                                                                                                                                                                                                
                                                                                                                                                                                                                  
  - DataContractUpdateTransitionAccessorsV0::data_contract() now returns Option<&DataContractInSerializationFormat> instead of &DataContractInSerializationFormat (V1 transitions don't embed the full contract)  
  - new_from_data_contract() method now always creates V0 transitions since V1 requires both old and new contracts. Use from_contract_update() for V1 delta-based transitions.                                    
  - Slightly different fee calculations for V1 transitions (27002430530 vs 27002504030 for V0)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [x] I have added "!" to the title and described breaking changes in the corresponding section if my code contains any
- [x] I have made corresponding changes to the documentation if needed

**For repository code-owners and collaborators only**
- [ ] I have assigned this pull request to a milestone


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

## Release Notes

* **New Features**
  * Added support for delta-based data contract updates enabling incremental changes instead of full contract replacement
  * Introduced validation methods for tokens, keywords, and schema definition updates
  * Added cost calculation for data contract registration and update operations
  * Extended platform versioning to support new state transition serialization formats

* **Bug Fixes & Error Handling**
  * Added specific error types for conflicting keywords and overlapping fields in contract updates
  * Improved validation of contract groups and token configurations

* **Tests**
  * Added comprehensive test suite for data contract update transitions and validation scenarios

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

### CODE REVIEW

File: packages/rs-drive/src/state_transition_action/contract/data_contract_update/v1/transformer.rs [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+        Ok(
+            validation_result.map(|new_data_contract| DataContractUpdateTransitionActionV1 {
+                old_data_contract,
+                data_contract: new_data_contract,
+                identity_contract_nonce: value.identity_contract_nonce,
+                user_fee_increase: value.user_fee_increase,
+            }),
+        )
+    }
```
Comment (thepastaclaw) [inline packages/rs-drive/src/state_transition_action/contract/data_contract_update/v1/transformer.rs:62]: **🔴 Blocking: Authorization bypass: any identity can update any contract via V1 path**

The V1 transformer fetches the old contract from state and builds the new contract with `owner_id: self.owner_id()` (the stored contract's real owner). This means `validate_update_v0`'s owner check (`self.owner_id() != new_data_contract.owner_id()`) always passes, since both sides are the stored owner.

Attack: craft a V1 update with `owner_id` = attacker's identity, `id` = victim's contract. Signature verification passes (attacker signs with their own keys). Nonce validation passes (attacker uses their own fresh nonce). `apply_update` rebuilds the new contract with the stored owner. `validate_update` compares stored_owner == stored_owner → passes. The update is applied to the victim's contract.

In V0, this works correctly because the new contract preserves the submitter's chosen owner_id from the embedded contract, so the owner check catches mismatches.

Fix: add an explicit check in `validate_state_v1` or `transform_into_action_v1` that compares the transition's `owner_id` against the fetched old contract's `owner_id` before proceeding.

<sub>source: ['codex-general', 'claude-security-auditor', 'codex-security-auditor']</sub>

<details>
<summary>🤖 Fix this with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `packages/rs-drive/src/state_transition_action/contract/data_contract_update/v1/transformer.rs`:
- [BLOCKING] lines 26-62: Authorization bypass: any identity can update any contract via V1 path
  The V1 transformer fetches the old contract from state and builds the new contract with `owner_id: self.owner_id()` (the stored contract's real owner). This means `validate_update_v0`'s owner check (`self.owner_id() != new_data_contract.owner_id()`) always passes, since both sides are the stored owner.

Attack: craft a V1 update with `owner_id` = attacker's identity, `id` = victim's contract. Signature verification passes (attacker signs with their own keys). Nonce validation passes (attacker uses their own fresh nonce). `apply_update` rebuilds the new contract with the stored owner. `validate_update` compares stored_owner == stored_owner → passes. The update is applied to the victim's contract.

In V0, this works correctly because the new contract preserves the submitter's chosen owner_id from the embedded contract, so the owner check catches mismatches.

Fix: add an explicit check in `validate_state_v1` or `transform_into_action_v1` that compares the transition's `owner_id` against the fetched old contract's `owner_id` before proceeding.
```
</details>

File: packages/rs-drive/src/state_transition_action/contract/data_contract_update/v1/transformer.rs [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+                false,
+                transaction,
+                platform_version,
+            )?
+            .1
+            .ok_or_else(|| {
+                Error::Protocol(Box::new(ProtocolError::DataContractNotPresentError(
+                    DataContractNotPresentError::new(value.id),
+                )))
+            })?
```
Comment (thepastaclaw) [inline packages/rs-drive/src/state_transition_action/contract/data_contract_update/v1/transformer.rs:39]: **🔴 Blocking: Non-consensus error for missing contract prevents nonce bump**

When the old contract doesn't exist, `try_from_transition` returns `Error::Protocol(Box::new(ProtocolError::DataContractNotPresentError(...)))`. In `transform_into_action_v1` (state/v1/mod.rs:456-468), the error handler checks `if let ProtocolError::ConsensusError(consensus_error) = *protocol_error` — which `DataContractNotPresentError` is NOT. So execution falls through to `Err(Error::Protocol(*protocol_error))`, returning an internal error instead of creating a `BumpIdentityDataContractNonceAction`.

Compare V0: in `validate_state_v0`, a missing contract is handled with a proper `BasicError::DataContractNotPresentError` consensus error wrapped in a bump action (state/v0/mod.rs:132-148).

Fix: either wrap the error as a consensus error in the transformer, or add explicit missing-contract handling in `transform_into_action_v1` before the generic error match.

<sub>source: ['claude-general']</sub>

<details>
<summary>🤖 Fix this with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `packages/rs-drive/src/state_transition_action/contract/data_contract_update/v1/transformer.rs`:
- [BLOCKING] lines 35-39: Non-consensus error for missing contract prevents nonce bump
  When the old contract doesn't exist, `try_from_transition` returns `Error::Protocol(Box::new(ProtocolError::DataContractNotPresentError(...)))`. In `transform_into_action_v1` (state/v1/mod.rs:456-468), the error handler checks `if let ProtocolError::ConsensusError(consensus_error) = *protocol_error` — which `DataContractNotPresentError` is NOT. So execution falls through to `Err(Error::Protocol(*protocol_error))`, returning an internal error instead of creating a `BumpIdentityDataContractNonceAction`.

Compare V0: in `validate_state_v0`, a missing contract is handled with a proper `BasicError::DataContractNotPresentError` consensus error wrapped in a bump action (state/v0/mod.rs:132-148).

Fix: either wrap the error as a consensus error in the transformer, or add explicit missing-contract handling in `transform_into_action_v1` before the generic error match.
```
</details>

File: packages/rs-drive/src/state_transition_action/contract/data_contract_update/v1/transformer.rs [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        // Fetch the old contract from Drive
+        let old_data_contract = drive
+            .get_contract_with_fetch_info_and_fee(
+                value.id.to_buffer(),
+                None,
+                false,
+                transaction,
+                platform_version,
+            )?
+            .1
```
Comment (thepastaclaw) [inline packages/rs-drive/src/state_transition_action/contract/data_contract_update/v1/transformer.rs:34]: **🟡 Suggestion: V1 update path discards fee for old contract state fetch**

The fee from `get_contract_with_fetch_info_and_fee()` is discarded via `.1` on the result tuple. In V0 (state/v0/mod.rs:122-129), this fee is properly charged to `execution_context`. Since the transformer doesn't have access to `execution_context`, the fee is lost. This means V1 contract updates are undercharged for the state read, which is a consensus-critical cost tracking issue per the review guidelines.

<sub>source: ['codex-general']</sub>

<details>
<summary>🤖 Fix this with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `packages/rs-drive/src/state_transition_action/contract/data_contract_update/v1/transformer.rs`:
- [SUGGESTION] lines 26-34: V1 update path discards fee for old contract state fetch
  The fee from `get_contract_with_fetch_info_and_fee()` is discarded via `.1` on the result tuple. In V0 (state/v0/mod.rs:122-129), this fee is properly charged to `execution_context`. Since the transformer doesn't have access to `execution_context`, the fee is lost. This means V1 contract updates are undercharged for the state read, which is a consensus-critical cost tracking issue per the review guidelines.
```
</details>

File: packages/rs-dpp/src/state_transition/state_transitions/contract/data_contract_create_transition/accessors/mod.rs [hunk 4]
```diff
@@ -10,12 +10,16 @@ impl DataContractCreateTransitionAccessorsV0 for DataContractCreateTransition {
     fn data_contract(&self) -> &DataContractInSerializationFormat {
         match self {
             DataContractCreateTransition::V0(transition) => &transition.data_contract,
+            DataContractCreateTransition::V1(_) => {
+                panic!("data_contract() accessor is not available for DataContractCreateTransitionV1 - use individual field accessors instead")
+            }
```
Comment (thepastaclaw) [inline packages/rs-dpp/src/state_transition/state_transitions/contract/data_contract_create_transition/accessors/mod.rs:15]: **🟡 Suggestion: data_contract() accessor panics on V1 create transitions**

The `DataContractCreateTransitionAccessorsV0` impl panics when `data_contract()` is called on V1 variants. While V1 transitions use individual field accessors, any generic code path that calls `transition.data_contract()` without version-checking will crash. The `set_data_contract()` method has the same issue.

The update transition handles this correctly by returning `Option` from its accessors. Consider matching that pattern, or ensure through type-level guarantees that V1 never reaches these code paths.

<sub>source: ['claude-general', 'claude-rust-quality']</sub>

<details>
<summary>🤖 Fix this with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `packages/rs-dpp/src/state_transition/state_transitions/contract/data_contract_create_transition/accessors/mod.rs`:
- [SUGGESTION] lines 13-15: data_contract() accessor panics on V1 create transitions
  The `DataContractCreateTransitionAccessorsV0` impl panics when `data_contract()` is called on V1 variants. While V1 transitions use individual field accessors, any generic code path that calls `transition.data_contract()` without version-checking will crash. The `set_data_contract()` method has the same issue.

The update transition handles this correctly by returning `Option` from its accessors. Consider matching that pattern, or ensure through type-level guarantees that V1 never reaches these code paths.
```
</details>

File: packages/rs-dpp/src/state_transition/state_transitions/contract/data_contract_update_transition/v1/mod.rs [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+            new_tokens,
+            remove_keywords,
+            add_keywords,
+            update_description,
+            identity_contract_nonce: identity_nonce,
+            user_fee_increase: 0,
+            signature_public_key_id: 0,
+            signature: Default::default(),
+        })
+    }
```
Comment (thepastaclaw) [inline packages/rs-dpp/src/state_transition/state_transitions/contract/data_contract_update_transition/v1/mod.rs:238]: **🟡 Suggestion: from_contract_update silently ignores document type deletions and group/token modifications**

The delta computation in `from_contract_update()` only iterates fields present in `new_contract`. If `old_contract` has document types, schema defs, groups, or tokens that are absent from `new_contract`, those removals are silently ignored — the resulting transition will not capture them. While the V1 struct intentionally only has `new_*`/`updated_*` fields (no `remove_*` for these categories), the method doesn't warn or error when it detects changes it cannot represent. A caller passing contracts with removed types would get a transition that doesn't fully represent their intent.

<sub>source: ['codex-security-auditor', 'codex-rust-quality', 'claude-rust-quality']</sub>

<details>
<summary>🤖 Fix this with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `packages/rs-dpp/src/state_transition/state_transitions/contract/data_contract_update_transition/v1/mod.rs`:
- [SUGGESTION] lines 132-238: from_contract_update silently ignores document type deletions and group/token modifications
  The delta computation in `from_contract_update()` only iterates fields present in `new_contract`. If `old_contract` has document types, schema defs, groups, or tokens that are absent from `new_contract`, those removals are silently ignored — the resulting transition will not capture them. While the V1 struct intentionally only has `new_*`/`updated_*` fields (no `remove_*` for these categories), the method doesn't warn or error when it detects changes it cannot represent. A caller passing contracts with removed types would get a transition that doesn't fully represent their intent.
```
</details>

File: packages/rs-drive-abci/src/execution/validation/state_transition/state_transitions/data_contract_update/state/v1/mod.rs [hunk 6]
```diff
@@ -0,0 +1,486 @@
...
+                    Ok(ConsensusValidationResult::new_with_data_and_errors(
+                        bump_action,
+                        validation_result.errors,
+                    ))
+                } else {
+                    Ok(validation_result.map(|update_action| update_action.into()))
+                }
+            }
+        }
+    }
```
Comment (thepastaclaw) [inline packages/rs-drive-abci/src/execution/validation/state_transition/state_transitions/data_contract_update/state/v1/mod.rs:485]: **🟡 Suggestion: Missing test coverage for V1 update edge cases**

No tests exist for: (1) V1 update where target contract does not exist (exercises the nonce-bump bug in finding #3), (2) V1 update where transition owner_id differs from stored contract owner_id (exercises the auth bypass in finding #1). These are consensus-critical paths that should have explicit test coverage.

<sub>source: ['claude-general']</sub>

<details>
<summary>🤖 Fix this with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `packages/rs-drive-abci/src/execution/validation/state_transition/state_transitions/data_contract_update/state/v1/mod.rs`:
- [SUGGESTION] lines 430-485: Missing test coverage for V1 update edge cases
  No tests exist for: (1) V1 update where target contract does not exist (exercises the nonce-bump bug in finding #3), (2) V1 update where transition owner_id differs from stored contract owner_id (exercises the auth bypass in finding #1). These are consensus-critical paths that should have explicit test coverage.
```
</details>

File: packages/rs-dpp/src/state_transition/state_transitions/contract/data_contract_update_transition/v1/v0_methods.rs [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+        let value = state_transition.signable_bytes()?;
+        let public_key =
+            identity
+                .loaded_public_keys
+                .get(&key_id)
+                .ok_or(ProtocolError::NonConsensusError(
+                    NonConsensusError::StateTransitionCreationError(
+                        "public key did not exist".to_string(),
+                    ),
+                ))?;
```
Comment (thepastaclaw) [inline packages/rs-dpp/src/state_transition/state_transitions/contract/data_contract_update_transition/v1/v0_methods.rs:61]: **💬 Nitpick: .ok_or in signing code returns generic error instead of specific key-not-found error**

The `loaded_public_keys.get(&key_id).ok_or(...)` pattern uses a generic `StateTransitionCreationError` string. Consider including the key_id in the error message for debuggability.

<sub>source: ['claude-rust-quality']</sub>

**Review by thepastaclaw — CHANGES_REQUESTED** [review]
## Code Review

This PR introduces V1 delta-based data contract update transitions for Protocol Version 12. The delta architecture is sound, but the V1 update path has a critical authorization bypass: the transition's owner_id is never verified against the stored contract's owner_id, allowing any identity to update any contract. Additionally, a copy-paste bug defaults to the wrong transition version, and a non-consensus error type for missing contracts prevents proper nonce bumping.

*Reviewed commit: 3afdbd2b*

🔴 3 blocking | 🟡 5 suggestion(s) | 💬 1 nitpick(s)

### 2 additional findings

<details>
<summary>🔴 <b>blocking:</b> Copy-paste bug: update transition defaults to create transition's version</summary>

`packages/rs-dpp/src/state_transition/state_transitions/contract/data_contract_update_transition/value_conversion.rs (lines 84-90)`

In `from_object()` (line 88) and `from_value_map()` (line 116), when no explicit version field is present, the code falls back to `contract_create_state_transition.default_current_version` instead of `contract_update_state_transition.default_current_version`. This means a V1 update transition deserialized without an explicit version field could be interpreted as the wrong version.

Both occurrences need `contract_create_state_transition` changed to `contract_update_state_transition`.

</details>
<details>
<summary>🟡 <b>suggestion:</b> new_from_data_contract dispatcher missing V1 match arm</summary>

`packages/rs-dpp/src/state_transition/state_transitions/contract/data_contract_create_transition/methods/mod.rs (lines 28-48)`

The factory dispatcher only routes to V0 (line 35) and returns `UnknownVersionError` for all other versions (line 44). The V1 implementation exists but is unreachable through this public entry point. If V1 create transitions are intended for Protocol Version 12, this dispatcher needs a V1 arm.

</details>

<details>
<summary>🤖 Prompt for all review comments with AI agents</summary>

```
These findings are from an automated code review. Verify each finding against the current code and only fix it if needed.

In `packages/rs-drive/src/state_transition_action/contract/data_contract_update/v1/transformer.rs`:
- [BLOCKING] lines 26-62: Authorization bypass: any identity can update any contract via V1 path
  The V1 transformer fetches the old contract from state and builds the new contract with `owner_id: self.owner_id()` (the stored contract's real owner). This means `validate_update_v0`'s owner check (`self.owner_id() != new_data_contract.owner_id()`) always passes, since both sides are the stored owner.

Attack: craft a V1 update with `owner_id` = attacker's identity, `id` = victim's contract. Signature verification passes (attacker signs with their own keys). Nonce validation passes (attacker uses their own fresh nonce). `apply_update` rebuilds the new contract with the stored owner. `validate_update` compares stored_owner == stored_owner → passes. The update is applied to the victim's contract.

In V0, this works correctly because the new contract preserves the submitter's chosen owner_id from the embedded contract, so the owner check catches mismatches.

Fix: add an explicit check in `validate_state_v1` or `transform_into_action_v1` that compares the transition's `owner_id` against the fetched old contract's `owner_id` before proceeding.
- [BLOCKING] lines 35-39: Non-consensus error for missing contract prevents nonce bump
  When the old contract doesn't exist, `try_from_transition` returns `Error::Protocol(Box::new(ProtocolError::DataContractNotPresentError(...)))`. In `transform_into_action_v1` (state/v1/mod.rs:456-468), the error handler checks `if let ProtocolError::ConsensusError(consensus_error) = *protocol_error` — which `DataContractNotPresentError` is NOT. So execution falls through to `Err(Error::Protocol(*protocol_error))`, returning an internal error instead of creating a `BumpIdentityDataContractNonceAction`.

Compare V0: in `validate_state_v0`, a missing contract is handled with a proper `BasicError::DataContractNotPresentError` consensus error wrapped in a bump action (state/v0/mod.rs:132-148).

Fix: either wrap the error as a consensus error in the transformer, or add explicit missing-contract handling in `transform_into_action_v1` before the generic error match.
- [SUGGESTION] lines 26-34: V1 update path discards fee for old contract state fetch
  The fee from `get_contract_with_fetch_info_and_fee()` is discarded via `.1` on the result tuple. In V0 (state/v0/mod.rs:122-129), this fee is properly charged to `execution_context`. Since the transformer doesn't have access to `execution_context`, the fee is lost. This means V1 contract updates are undercharged for the state read, which is a consensus-critical cost tracking issue per the review guidelines.

In `packages/rs-dpp/src/state_transition/state_transitions/contract/data_contract_update_transition/value_conversion.rs`:
- [BLOCKING] lines 84-90: Copy-paste bug: update transition defaults to create transition's version
  In `from_object()` (line 88) and `from_value_map()` (line 116), when no explicit version field is present, the code falls back to `contract_create_state_transition.default_current_version` instead of `contract_update_state_transition.default_current_version`. This means a V1 update transition deserialized without an explicit version field could be interpreted as the wrong version.

Both occurrences need `contract_create_state_transition` changed to `contract_update_state_transition`.

In `packages/rs-dpp/src/state_transition/state_transitions/contract/data_contract_create_transition/accessors/mod.rs`:
- [SUGGESTION] lines 13-15: data_contract() accessor panics on V1 create transitions
  The `DataContractCreateTransitionAccessorsV0` impl panics when `data_contract()` is called on V1 variants. While V1 transitions use individual field accessors, any generic code path that calls `transition.data_contract()` without version-checking will crash. The `set_data_contract()` method has the same issue.

The update transition handles this correctly by returning `Option` from its accessors. Consider matching that pattern, or ensure through type-level guarantees that V1 never reaches these code paths.

In `packages/rs-dpp/src/state_transition/state_transitions/contract/data_contract_create_transition/methods/mod.rs`:
- [SUGGESTION] lines 28-48: new_from_data_contract dispatcher missing V1 match arm
  The factory dispatcher only routes to V0 (line 35) and returns `UnknownVersionError` for all other versions (line 44). The V1 implementation exists but is unreachable through this public entry point. If V1 create transitions are intended for Protocol Version 12, this dispatcher needs a V1 arm.

In `packages/rs-dpp/src/state_transition/state_transitions/contract/data_contract_update_transition/v1/mod.rs`:
- [SUGGESTION] lines 132-238: from_contract_update silently ignores document type deletions and group/token modifications
  The delta computation in `from_contract_update()` only iterates fields present in `new_contract`. If `old_contract` has document types, schema defs, groups, or tokens that are absent from `new_contract`, those removals are silently ignored — the resulting transition will not capture them. While the V1 struct intentionally only has `new_*`/`updated_*` fields (no `remove_*` for these categories), the method doesn't warn or error when it detects changes it cannot represent. A caller passing contracts with removed types would get a transition that doesn't fully represent their intent.

In `packages/rs-drive-abci/src/execution/validation/state_transition/state_transitions/data_contract_update/state/v1/mod.rs`:
- [SUGGESTION] lines 430-485: Missing test coverage for V1 update edge cases
  No tests exist for: (1) V1 update where target contract does not exist (exercises the nonce-bump bug in finding #3), (2) V1 update where transition owner_id differs from stored contract owner_id (exercises the auth bypass in finding #1). These are consensus-critical paths that should have explicit test coverage.
```
</details>

**Comment by thepastaclaw** [comment]
🕓 Review not started yet because this PR is a draft.

- [ ] **Request normal review** — click when the PR is ready for review.
- [ ] **Request priority review** — click to move this review to the front of the queue.

_Commit 3afdbd2b. Normal review starts when eligible; priority review starts as soon as a slot is available._

**Comment by PastaPastaPasta** [comment]
The V1 data-contract transition work in this draft has already landed through subsequent commits and is present in the current repository, including V1 create/update transition modules, validation, Drive handling, and protocol-version wiring. This draft is therefore superseded and no longer represents the implementation to review. I’m closing it; follow-up gaps should be filed against the current code.

---
🤖 Posted autonomously by Codex on behalf of pasta.
